src/rust/scheduler/coop_yield.rs

```diff
@@ -0,0 +1,89 @@
+// Imports
+//======================================================================================================================
+
+use ::std::{
+    future::Future,
+    pin::Pin,
+    task::{
+        Context,
+        Poll,
+    },
+};
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// Yield is a future that lets the currently running coroutine cooperatively yield because it cannot make progress.
+/// Coroutines are expected to use the standalone async functions to create yield points.
+struct Yield {
+    /// Is this yield complete? In general, the yield becomes runnable the second time that it is polled because it was
+    /// woken.
+    runnable: bool,
+    /// Should we yield until a wake signal or just one quantum? In general, we should wait until woken but sometimes
+    /// we want to immediately poll the coroutine again.
+    once: bool,
+}
+
+//======================================================================================================================
+// Associate Functions
+//======================================================================================================================
+
+impl Yield {
+    pub fn new(once: bool) -> Self {
+        Self { runnable: false, once }
+    }
+
+    pub fn is_runnable(&self) -> bool {
+        self.runnable
+    }
+
+    pub fn set_runnable(&mut self) {
+        self.runnable = true;
+    }
+
+    pub fn yield_once(&self) -> bool {
+        self.once
+    }
+}
+
+//======================================================================================================================
+// Trait Implementations
+//======================================================================================================================
+
+impl Future for Yield {
+    type Output = ();
+
+    /// Polls the underlying accept operation.
+    fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
+        let self_: &mut Yield = self.get_mut();
+
+        if self_.is_runnable() {
+            return Poll::Ready(());
+        }
+
+        self_.set_runnable();
+        // If we are only yielding for one scheduler tick, then schedule immediately. Otherwise, wait for another
+        // coroutine to send a signal via the waker.
+        if self_.yield_once() {
+            context.waker().wake_by_ref();
+        }
+        Poll::Pending
+    }
+}
+
+//==============================================================================
+// Standalone Functions
+//==============================================================================
+
+pub async fn yield_once() {
+    Yield::new(true).await
+}
+
+pub async fn yield_until_wake() {
+    Yield::new(false).await
+}
```

Comment:
Could we make this private associated functions of `Yield`?

Reply:
These can't be private because these are the ones that the libOSes called. I can make the Yield associated function private though.